ONSdigital/dp-import-api
# Support Cantabular data type and Kafka producer

### What

Modified import API to support Cantabular import jobs, according to trello card https://trello.com/c/nNlJKQfM/5115-implement-changes-for-dp-import-api-for-cantabular-data-type

- Created Cantabular Kafka producer, with topic `cantabular-dataset-instance-started`
  - Topic is configurable
  - Added checker, logging, and graceful shutdown for new Kafka producer in service.go
  - Provided new kafka producer to ImporQueue
- ImportQueue sends a message to the new kafka producer when a job with 'cantabular' format is received
  - Used new cantabular kafka message schema, creared in `dp-import`
  - validation bug for V4 has been fixed (the validation was failing if len(files) **and** len(instances) were both different than 1. Now it fails if either one of them (**or**) is different than 1)
- Improved testing for importQueue

### How to review

- Make sure code changes make sense
- Make sure unit tests pass
- Run `make debug`, do an HTTP health request: `http://localhost:21800/health` and validate that a health-check checker appears for `Kafka Cantabular Dataset Instance Started Producer`
  - Check that graceful shutdown works as expected, without error, when you kill the service (you should see the following log messages: `closing cantabular dataset instance started producer`, `closing initialised kafka producer` and `successfully closed kafka producer`)

(Info only, no need to reproduce, but feel free to do it).
I tested the functionality locally by doing the following:
- [0] (prep): run the following:
  - dp-import-api (with this branch)
  - dp-dataset-api
  - dp-recipe-api
  - zebedee
  - dp-kafka consumer example, listenint to the cantabular topic (`export KAFKA_CONSUMED_TOPIC=cantabular-dataset-instance-started`)
- [1] Manually create a recipe in my local MongoDB by copying an existing V4 recipe and changing the following:
  - `"format" : "cantabular_blob"` (repeat for `cantabular_table`)
  - files is an empty list `"files" : []`
- [2] Post a new Import job: `POST http://localhost:21800/jobs`
  - body contains the recipe Id created in step 1: `{"recipe" : "<cantabular_recipe_id>"}`
  - Should return `201 CREATED`
- [3] Change the state to 'submitted' by using the put endpoint, with the new job id from step 2: `PUT http://localhost:21800/jobs/<cantabular_job_id>`
  - should return `200 OK`
- [4] Check that a kafka message was produced with topic `cantabular-dataset-instance-started`, by checking the dp-kafka consumer that was run in step 0.

### Who can review

Anyone
